tools/cldr-code/src/main/java/org/unicode/cldr/util/personname/PersonNameFormatter.java

```diff
@@ -522,28 +619,96 @@ public NamePattern getBestMatch(NameObject nameObject, FormatParameters nameForm
-            // * no ParameterMatcher should be completely masked by previous ones
-            // * each entry in ParameterMatcher must have at least one NamePattern
-            // * the final entry must have MATCH_ALL as the key
+
+            ParameterMatcher lastKey = null;
+            Set<FormatParameters> remaining = new LinkedHashSet<>(FormatParameters.all());
+
+            for (Entry<ParameterMatcher, Collection<NamePattern>> entry : formatParametersToNamePattern.asMap().entrySet()) {
+                ParameterMatcher key = entry.getKey();
+                Collection<NamePattern> values = entry.getValue();
+
+                // TODO Mark No ParameterMatcher should be completely masked by any previous ones
+                // Except for the final MATCH_ALL
+                // The following code starts with a list of all the items, and removes any that match
+                int matchCount = 0;
+                int matchAllCount = 0;
+                for (Iterator<FormatParameters> rest = remaining.iterator(); rest.hasNext(); ) {
+                    FormatParameters item = rest.next();
+                    if (key.matches(item)) {
+                        rest.remove();
+                        ++matchCount;
+                        if (DEBUG) {
+                            System.out.println(" * " + item + " matches " + key);
+                        }
+                    }
+                }
+                if (matchCount == 0) {
+                    throw new IllegalArgumentException("key is masked by previous values: " + key);
+                }
+
+                // Each entry in ParameterMatcher must have at least one NamePattern
+                if (values.isEmpty()) {
+                    throw new IllegalArgumentException("key has no values: " + key);
+                }
+                lastKey = key;
+            }
+
+            // The final entry must have MATCH_ALL as the key
+            if (!lastKey.equals(ParameterMatcher.MATCH_ALL)) {
+                throw new IllegalArgumentException("last key is not MATCH_ALL" + lastKey);
+            }
+        }
+
+        /**
+         * Build from strings for ease of testing
+         * TODO Mark make the localeToOrder from strings also
+         */
+        public NamePatternData(ImmutableMap<ULocale, Order> localeToOrder, String...formatParametersToNamePatterns ) {
+            this(localeToOrder, parseFormatParametersToNamePatterns(formatParametersToNamePatterns));
         }
 
-        // TODO Mark add method that takes a string instead of a NamePatten
+        private static ImmutableListMultimap<ParameterMatcher, NamePattern> parseFormatParametersToNamePatterns(String... formatParametersToNamePatterns) {
+            int count = formatParametersToNamePatterns.length;
+            if ((count % 2) != 0) {
+                throw new IllegalArgumentException("Must have even number of strings, fields => pattern: " + Arrays.asList(formatParametersToNamePatterns));
```

Comment:
Would be nice to have a comment explaining why this is... Can be for next iteration.